src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -0,0 +1,444 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowEndNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
+
+/**
+ * Scanner that will scan down all forks when we hit parallel blocks before continuing, but generally runs in linear order
+ * <p/>Think of it as the opposite of {@link DepthFirstScanner}.
+ *
+ * <p/>This is a fairly efficient way to visit all FlowNodes, and provides three useful guarantees:
+ * <ul>
+ *   <li>Every FlowNode is visited, and visited EXACTLY ONCE (not true for LinearScanner)</li>
+ *   <li>All parallel branches are visited before we move past the parallel block (not true for DepthFirstScanner)</li>
+ *   <li>For EVERY block, the BlockEndNode is visited before the BlockStartNode (not true for DepthFirstScanner, with parallels)</li>
+ * </ul>
+ *
+ * <p/>The big advantages of this approach:
+ * <ul>
+ *     <li>Blocks are visited in the order they end (no backtracking) - helps with working a block at a time</li>
+ *     <li>Points are visited in linear order within a block (easy to use for analysis)</li>
+ *     <li>Minimal state information needed</li>
+ *     <li>Branch information is available for use here</li>
+ * </ul>
+ *
+ * @author <samvanoort@gmail.com>Sam Van Oort</samvanoort@gmail.com>
+ */
+public class ForkScanner extends AbstractFlowScanner {
+
+    // Last element in stack is end of myCurrent parallel start, first is myCurrent start
+    ArrayDeque<ParallelBlockStart> parallelBlockStartStack = new ArrayDeque<ParallelBlockStart>();
+
+    /** FlowNode that will terminate the myCurrent parallel block */
+    FlowNode currentParallelStartNode = null;
+
+    ParallelBlockStart currentParallelStart = null;
+
+    private boolean walkingFromFinish = false;
+
+    @Override
+    protected void reset() {
+        parallelBlockStartStack.clear();
+        currentParallelStart = null;
+        currentParallelStartNode = null;
+        myCurrent = null;
+        myNext = null;
+    }
+
+    /** If true, we are walking from the flow end node and have a complete view of the flow */
+    public boolean isWalkingFromFinish() {
```

Comment:
There are implications when you're dealing with an unfinished flow (incomplete blocks, for example, potentially unclosed parallels).